### PR TITLE
common: avoid pthread_mutex_unlock twice

### PR DESCRIPTION
**Pull request by Zhiwei-Dai**
Before the pr #36942, ret of o->reclaim() is always true.
Till now, lru lane lock maybe unlock twice. The pr fix it to avoid undefined behaviors.

Signed-off-by: Dai Zhiwei <daizhiwei3@huawei.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

### CODE REVIEW

File: src/common/cohort_lru.h [hunk 1]
```diff
@@ -159,10 +159,10 @@ namespace cohort {
 	      lane.lock.unlock();
 	      return o;
 	    } else {
-	      // XXX can't make unreachable (means what?)
 	      --(o->lru_refcnt);
 	      o->lru_flags &= ~FLAG_EVICTING;
-	      /* unlock in next block */
+	      /* avoid unlock twice */
+	      continue;
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/cohort_lru.h:165]: IIUC, `unique_lock` should relinquish the held lock when it is destroyed. so your concern should already have been addressed by my change. am i missing anything?

[the remark this answers is not in the document]
Comment (Zhiwei-Dai) [reply src/common/cohort_lru.h:165]: Agree

File: src/common/cohort_lru.h [hunk 2]
```diff
@@ -139,15 +139,15 @@ namespace cohort {
 	for (int ix = 0; ix < n_lanes; ++ix,
 	       lane_ix = next_evict_lane()) {
 	  Lane& lane = qlane[lane_ix];
-	  lane.lock.lock();
+	  std::unique_lock<std::mutex> lane_lock{lane.lock};
```
Comment (tchaikov) [inline src/common/cohort_lru.h:142]: I'd prefer not specifying the exact type of `lane_lock`. as, the type of `Lane::LK` is template parameter of `LRU`. and C++17 is able to deduce the type of `lane_lock` even we don't specify it, so following code should compile and work:
```c++
std::unique_lock lane_lock{lane.lock};
```

**Comment by tchaikov** [comment]
@Zhiwei-Dai please see my comment.

File: src/common/cohort_lru.h (hunk 2 above)
Comment (Zhiwei-Dai) [reply src/common/cohort_lru.h:142]: Well. pass complication and testcases

Comment (Zhiwei-Dai) [reply src/common/cohort_lru.h]: recommit it

**Review by mattbenjamin — APPROVED** [review]
Yes, we could unlock twice because in the !reclaim(...) case there is no early return.  I do think the current logic using std::unique_lock is sound

**Comment by cbodley** [comment]
passed qa https://pulpito.ceph.com/cbodley-2021-11-12_16:45:53-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**